src/Float16.h

```diff
@@ -124,4 +129,127 @@ HALIDE_ALWAYS_INLINE halide_type_t halide_type_of<Halide::float16_t>() {
     return halide_type_t(halide_type_float, 16);
 }
 
+namespace Halide {
+
+/** Class that provides a type that implements half precision
+ *  floating point using the bfloat16 format.
+ *
+ *  This type is enforced to be 16-bits wide and maintains no state
+ *  other than the raw bits so that it can passed to code that checks
+ *  a type's size and used for buffer_t allocation. */
+struct bfloat16_t {
+
+    static const int mantissa_bits = 7;
+    static const uint16_t sign_mask = 0x8000;
+    static const uint16_t exponent_mask = 0x7f80;
+    static const uint16_t mantissa_mask = 0x007f;
+
+    /// \name Constructors
+    /// @{
+
+    /** Construct from a float, double, or int using
+     * round-to-nearest-ties-to-even. Out-of-range values become +/-
+     * infinity.
+     */
+    // @{
+    bfloat16_t(float value);
```

Comment:
are we sure we don't want `explicit` here?

Reply:
It's to support things like some_bfloat16 > 0, or some_bfloat16 + 1, but now that you mention it it's going to have unwanted side-effects, like some_bfloat16 + 1.0f won't get promoted to float.

I guess I could add more overloads below instead. Do you think that would be better?

Reply:
Yeah, non-explicit single-arg ctors that take 'primitive' types (numbers, strings) are a bug waiting to happen; IMHO we'd be better off making them all explicit and seeing how much pain that really is in practice. Before adding more overloads it may be worth just seeing how many explicit casts are necessary; if it's small-ish, I think that's the better option.

Reply:
Done. It was smallish.